aptos-move/move-examples/nfts_as_accounts/sources/token.move

```diff
@@ -0,0 +1,200 @@
+///       the token to mutate it as appropriate.
+///
+/// Some notes:
+/// * This module doesn't have clean error handling because some operations are not cheap. It is
+///   more effective to just let the upstream abort happen, rather than try to handle it locally,
+///   unless there's a means to recover.
+///
+/// TODO(@davidiw):
+/// * add appropriate set of accessors for fields
+/// * consider tracking ownership if the token is inserted into a TokenStore
+/// * create_token should probably have another function that returns the signer for descendants
+module nfts_as_accounts::token {
+    use std::signer;
+    use std::string::{Self, String};
+    use std::vector;
+
+    use aptos_framework::account;
+
+    const ENOT_THE_CREATOR: u64 = 1;
+
+    /// Represents the common fields to all tokens.
+    struct Token has key {
+        /// An optional categorization of similar token, there are no constraints on collections.
+        collection: String,
+        /// The original creator of this token.
+        creator: address,
+        /// A brief description of the token.
+        description: String,
+        /// Determines which fields are mutable.
+        mutability_config: MutabilityConfig,
+        /// The name of the token, which should be unique within the collection; the length of name
+        ///should be smaller than 128, characters, eg: "Aptos Animal #1234"
+        name: String,
+        /// The denominator and numerator for calculating the royalty fee; it also contains payee
+        /// account address for depositing the Royalty
+        royalty: Royalty,
+        /// The Uniform Resource Identifier (uri) pointing to the JSON file stored in off-chain
+        /// storage; the URL length will likely need a maximum any suggestions?
+        uri: String,
+    }
+
+    /// This config specifies which fields in the TokenData are mutable
+    struct MutabilityConfig has copy, drop, store {
+        description: bool,
+        name: bool,
+        uri: bool,
+    }
+
+    /// The royalty of a token
+    struct Royalty has copy, drop, store {
+        numerator: u64,
+        denominator: u64,
+        /// The recipient of royalty payments. See the `shared_account` for how to handle multiple
+        /// creators.
+        payee_address: address,
+    }
+
+    struct TokenRef has store {
+        inner: account::SignerCapability,
+    }
+
+    public fun create_token(
+        creator: &signer, 
+        collection: String,
+        description: String,
+        mutability_config: MutabilityConfig,
+        name: String,
+        royalty: Royalty,
+        uri: String,
+    ): TokenRef {
+        let seed = *string::bytes(&collection);
+        vector::append(&mut seed, b"::");
+        vector::append(&mut seed, *string::bytes(&name));
+        // To keep costs down, this function does not check to see if the token already exists
+        let (token_account, inner) = account::create_resource_account(creator, seed);
```

Comment:
@msmouse would really hate this

Reply:
I'd love to be able to say that the entirety of one of these entities is stored in a single slot -- inclusive of the account data.

Reply:
Actually, we could probably add something new to account, like `create_object`. It wouldn't let an account exist at that resource but would give you similar signer capabilities like resource accounts. That would get rid of one slot, but at the cost of eliminating the TokenRef and everything just being an ObjectRef and no guaranteed to exist Token struct... worth chewing on more, but that's a framework change :D.

Reply:
When I was thinking of achieving composability in Aptos, manipulating text was a viable solution.

Reply:
What do you mean?

Reply:
`
character: string = 'gender:male;sword_type:1;'
`
like this

Reply:
cc @msmouse lite accounts please.

Reply:
Definitely don't agree @movekevin, in my opinion storage needs to evolve at this point. Considering that both resources could be theoretically cleaned up and the cost returned to the user, there's no point in that type of over optimization.

Reply:
> Definitely don't agree @movekevin, in my opinion storage needs to evolve at this point.
What do you mean?
Wouldn't "lite accounts" or "object accounts" be an evolution of storage already?

Reply:
oh, got ya... incoming...

Reply:
> entirety of one of these entities is stored in a single slot -- inclusive of the account data.
Could you elaborate this?